src/fitting.jl

```diff
@@ -192,3 +192,88 @@ function fit_impl(
 
   γ, ϵ
 end
+
+function fit_impl(
+  V::Type{<:PowerVariogram},
+  g::EmpiricalVariogram,
+  algo::WeightedLeastSquares;
+  scale=nothing,
+  exponent=nothing,
+  nugget=nothing,
+  maxscale=nothing,
+  maxexponent=nothing,
+  maxnugget=nothing
+)
+  # coordinates of empirical variogram
+  x = g.abscissas
+  y = g.ordinates
+  n = g.counts
+
+  # discard invalid bins
+  x = x[n .> 0]
+  y = y[n .> 0]
+  n = n[n .> 0]
+
+  # strip units of coordinates
+  ux = unit(eltype(x))
+  uy = unit(eltype(y))
+  x′ = ustrip.(x)
+  y′ = ustrip.(y)
+
+  # strip units of kwargs
+  scale′ = isnothing(scale) ? scale : ustrip(ux, scale)
+  exponent′ = isnothing(exponent) ? exponent : ustrip(uy, exponent)
+  nugget′ = isnothing(nugget) ? nugget : ustrip(uy, nugget)
+  maxscale′ = isnothing(maxscale) ? maxscale : ustrip(ux, maxscale)
+  maxexponent′ = isnothing(maxexponent) ? maxexponent : ustrip(uy, maxexponent)
+  maxnugget′ = isnothing(maxnugget) ? maxnugget : ustrip(uy, maxnugget)
+
+  # evaluate weights
+  f = algo.weightfun
+  w = isnothing(f) ? n / sum(n) : map(xᵢ -> ustrip(f(xᵢ)), x)
+
+  # objective function
+  function J(θ)
+    γ = V(scaling=θ[1], nugget=θ[2], exponent=θ[3])
+    sum(i -> w[i] * (γ(x′[i]) - y′[i])^2, eachindex(w, x′, y′))
+  end
+
+  # linear constraints
+  # 1. scaling ≥ 0
+  # 2. 0 ≤ exponent ≤ 2
+  L(θ) = [θ[1] ≥ 0.0 ? 0.0 : -θ[1], θ[3] ≥ 0.0 ? 0.0 : -θ[3], 2.0 ≥ θ[3] ? 0.0 : θ[3] - 2.0]
+
+  # penalty for linear constraint (J + λL)
+  λ = fill(sum(yᵢ -> yᵢ^2, y′), 3)
```

Comment:
@stepanoslejsek one more thing... Maybe we should refactor this expression in terms of L_1, L_2 and L_3 so that L(theta) = L_1(theta) + L_2(theta) + L_3(theta). You can then define \lambda as a scalar below and avoid unnecessary vector allocations. What do you think?

Reply:
Sure, that would be good suggestion, I agree to do so.